### PR TITLE
Refactor PayPal Module

### PR DESCRIPTION
**Pull request by hirale**
### Description (*)

Refactor PayPal module to meet PayPal REST APIs with https://github.com/paypal/PayPal-PHP-Server-SDK

### Related Pull Requests

- see OpenMage/magento-lts#4790

## ✅ Current Status

The standard purchase flow has been fully tested and works as expected.  
It supports:

- **Refund**
- **Void**
- **Authorize**
- **Capture**

For **Authorize** transactions, I have created a cronjob to process automatic re-authorize after 3 days honor period.
You can find more details here:  
👉 [PayPal Checkout Payment Intent Documentation](https://developer.paypal.com/serversdk/typescript/models/enumerations/checkout-payment-intent/)

---

## 🔜 Next Steps (Help Needed)

I’ll need help with the following:

1. **Recurring Payments**
   - Development and testing of recurring payment functionality.

2. **Compatibility with One Step Checkout Extensions**
   - Review and ensure compatibility with popular one-step checkout modules.

3. **Code Testing**
   - Improve test coverage.
   - Identify critical areas needing unit or integration tests.
   - Validate edge cases and PayPal SDK integration scenarios.

4. **Backwards Compatibility**
   - Ensure smooth backward compatibility where possible.
   - The current implementation **removes all old PayPal payment methods from the install scripts**, keeping only the corresponding **payment title config**.
   - Review upgrade paths for existing merchants migrating from previous PayPal integration to this new module version.

---

## 📦 Versioning Notes

- The module version has been bumped directly to `2.0.0.0`.
- All previous PayPal payment methods and install scripts have been removed.
- Only config-based **payment titles** remain for legacy visibility where needed.

---

## 📚 Reference

- [PayPal Checkout API Documentation](https://developer.paypal.com/api/rest/)

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by empiricompany** [comment]
@hirale, I just checked out the PR and then ran a composer update. The first issue I’m encountering is that the installation script isn’t being launched because there’s already a resource in the database:  
paypal_setup  1.6.0.6  1.6.0.6  

Maybe we should keep the previous migrations and rename it to upgrade-1.6.0.6-2.0.0.0.php?

File: app/code/core/Mage/Paypal/controllers/PaymentController.php [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+            ->setCustomerGroupId(Mage_Customer_Model_Group::NOT_LOGGED_IN_ID);
+        return $this;
+    }
+
+    /**
+     * Prepares the quote for a new customer registration during checkout.
+     */
+    protected function _prepareNewCustomerQuote(): self
+    {
+        $quote      = $this->_quote;
```
Comment (empiricompany) [inline app/code/core/Mage/Paypal/controllers/PaymentController.php]: When placing an order, I'm encountering this issue:

```suggestion
        $quote      = $this->_getQuote();
```
https://github-openmage.loc/paypal/payment/placeOrder/

Fatal error: Uncaught Error: Call to a member function getBillingAddress() on bool in /shared/httpd/github-openmage/htdocs/app/code/core/Mage/Paypal/controllers/PaymentController.php:303
Stack trace: 
#0 /shared/httpd/github-openmage/htdocs/app/code/core/Mage/Paypal/controllers/PaymentController.php(152): Mage_Paypal_PaymentController->_prepareNewCustomerQuote() 
#1 /shared/httpd/github-openmage/htdocs/app/code/core/Mage/Core/Controller/Varien/Action.php(410): Mage_Paypal_PaymentController->placeOrderAction() 
#2 /shared/httpd/github-openmage/htdocs/app/code/core/Mage/Core/Controller/Varien/Router/Standard.php(256): Mage_Core_Controller_Varien_Action->dispatch('placeOrder') 
#3 /shared/httpd/github-openmage/htdocs/app/code/core/Mage/Core/Controller/Varien/Front.php(174): Mage_Core_Controller_Varien_Router_Standard->match(Object(Mage_Core_Controller_Request_Http)) 
#4 /shared/httpd/github-openmage/htdocs/app/code/core/Mage/Core/Model/App.php(349): Mage_Core_Controller_Varien_Front->dispatch() 
#5 /shared/httpd/github-openmage/htdocs/app/Mage.php(743): Mage_Core_Model_App->run(Array) 
#6 /shared/httpd/github-openmage/htdocs/index.php(50): Mage::run('', 'store') 
#7 {main} thrown in /shared/httpd/github-openmage/htdocs/app/code/core/Mage/Paypal/controllers/PaymentController.php on line 303

**Comment by empiricompany** [comment]
**Refund functionality:**  
If I create a Credit Memo from the order, I only see the "Refund Offline" button.  
However, if I do it from the invoice, I correctly see two buttons: "Refund Offline" and "Refund."

**Comment by empiricompany** [comment]
> **Refund functionality:** If I create a Credit Memo from the order, I only see the "Refund Offline" button. However, if I do it from the invoice, I correctly see two buttons: "Refund Offline" and "Refund."

Sorry, I was mistaken. This is the correct behavior because multiple invoices can be linked to an order, and online refunds require referencing a specific invoice/transaction. Let’s ignore this as it’s unrelated to this PR. A separate issue can be opened if needed to review this behavior further.

File: app/code/core/Mage/Paypal/controllers/PaymentController.php (hunk 1 above)
Comment (empiricompany) [reply app/code/core/Mage/Paypal/controllers/PaymentController.php]: Perhaps it would be better to replace all occurrences of `$this->_quote` with the method `$this->_getQuote()`?

**Comment by hirale** [comment]
> @hirale, I just checked out the PR and then ran a composer update. The first issue I’m encountering is that the installation script isn’t being launched because there’s already a resource in the database: paypal_setup 1.6.0.6 1.6.0.6
> 
> Maybe we should keep the previous migrations and rename it to upgrade-1.6.0.6-2.0.0.0.php?

You are right, i didn't try a fresh install.

**Comment by empiricompany** [comment]
Card errors need to be handled.
https://developer.paypal.com/tools/sandbox/card-testing/#link-simulatecarderrorscenarios

Currently for this scenarios, the checkout is stuck in the processing phase.

**Comment by empiricompany** [comment]
I encountered the following issue:  
When I reach step 5 - Order Review in the checkout process, the buttons load correctly. However, if I go back to step 4 and then proceed again, no event is triggered, and the "PLACE ORDER" button is not replaced.
P.S. If, instead, I load step 3 and proceed, it works correctly.

**Comment by hirale** [comment]
> I encountered the following issue: When I reach step 5 - Order Review in the checkout process, the buttons load correctly. However, if I go back to step 4 and then proceed again, no event is triggered, and the "PLACE ORDER" button is not replaced. P.S. If, instead, I load step 3 and proceed, it works correctly.

I cannot reproduce. Can you record a video with network panel? Thanks.

https://github.com/user-attachments/assets/67a7bb4e-3554-4e96-844e-bf236c2be560

**Comment by empiricompany** [comment]
> I cannot reproduce. Can you record a video with network panel? Thanks.
> Video.mp4

@hirale To reproduce the issue, you _must not_ switch the payment method. Simply go back to step 4, leave PayPal selected, and proceed.

**Comment by hirale** [comment]
> > I cannot reproduce. Can you record a video with network panel? Thanks.
> > Video.mp4
> 
> @hirale To reproduce the issue, you _must not_ switch the payment method. Simply go back to step 4, leave PayPal selected, and proceed.

Fixed.

**Comment by empiricompany** [comment]
There are issues with the MyBank checkout. This is a successful payment:
mybank

**Comment by hirale** [comment]
> Card errors need to be handled. https://developer.paypal.com/tools/sandbox/card-testing/#link-simulatecarderrorscenarios
> 
> Currently for this scenarios, the checkout is stuck in the processing phase.

```
[result:Core\Types\Sdk\CoreApiResponse:private] => Array
        (
            [name] => UNPROCESSABLE_ENTITY
            [details] => Array
                (
                    [0] => stdClass Object
                        (
                            [issue] => INSTRUMENT_DECLINED
                            [description] => The instrument presented  was either declined by the processor or bank, or it can't be used for this payment.
                        )

                )

            [message] => The requested action could not be performed, semantically incorrect, or failed business validation.
            [debug_id] => f205275aef26b
            [links] => Array
                (
                    [0] => stdClass Object
                        (
                            [href] => https://developer.paypal.com/api/rest/reference/orders/v2/errors/#INSTRUMENT_DECLINED
                            [rel] => information_link
                            [method] => GET
                        )

                    [1] => stdClass Object
                        (
                            [href] => https://www.sandbox.paypal.com/checkoutnow\?token=2GE05246XH504793L
                            [rel] => redirect
                            [method] => GET
                        )

                )

        )
```

I have tried below card triggers, they return same error. So i added an error message after PayPal button, customer can retry without problems.

- CCREJECT-REFUSED	
- CCREJECT-SF	
- CCREJECT-EC	
- CCREJECT-IRC	
- CCREJECT-IA	
- CCREJECT-BANK_ERROR	


image

**Comment by hirale** [comment]
> There are issues with the MyBank checkout. This is a successful payment: mybank

Regarding MyBank and other payment sources — there's a bug in the SDK ([PayPal-PHP-Server-SDK issue #46](https://github.com/paypal/PayPal-PHP-Server-SDK/issues/46)). To work around this, I’ve implemented a temporary fix to keep everything functioning properly.

We need to test also other payment methods listed here.
https://developer.paypal.com/docs/checkout/apm/#available-payment-methods

**Comment by sreichel** [comment]
👍
